### DISCUSSION TITLE
How to calculate the momentum flux?

### DISCUSSION THREAD

**Discussion by Tinydog8** [CliMA/Oceananigans.jl#3053] [Computational science!]
Hi all!

I'm trying to calculate the momentum flux term (both resolved and sgs), but the result seems strange near the bounded grids.
Here is the initial data of u'w' (Center)and νₑ∂z(u) (z Face). Should I replace the bounded data with my forcing? (Such as Qᵘ)

```julia
u_bcs = FieldBoundaryConditions(top=FluxBoundaryCondition(Qᵘ), bottom = drag_bc_u)
v_bcs = FieldBoundaryConditions(bottom = drag_bc_v)
```

![微信图片_20230406095328](https://user-images.githubusercontent.com/117068415/230252677-d62e5bc2-e698-4272-af44-719a6401ffad.png)

Thanks!

**Comment by glwagner** [comment]
Hi! It depends on what the boundary conditions are, so can you post more of the code? For example, what is `drag_bc_u` and `drag_bc_v`?

That said we are lacking a general utility for this, and I think we need one as it would be useful to a lot of people. 

EDIT: I may have been confused. Are you trying to calculate u'w' or are you trying to calculate the momentum flux at the boundary?

**Reply by glwagner** [reply]
Can you clarify how you are calculating the momentum flux now, what you expect to see, and what is the issue with the plots?

**Reply by Tinydog8** [reply]
> Hi! It depends on what the boundary conditions are, so can you post more of the code? For example, what is `drag_bc_u` and `drag_bc_v`?
> 
> That said we are lacking a general utility for this, and I think we need one as it would be useful to a lot of people.
> 
> EDIT: I may have been confused. Are you trying to calculate u'w' or are you trying to calculate the momentum flux at the boundary?

```julia
const u₁₀=8  # m s⁻¹, average wind velocity 10 meters above the ocean`
const cᴰʷ=(0.75+0.067*u₁₀)*1e-3 # dimensionless drag coefficient`
const ρₐ=1.225  # kg m⁻³, average density of air at sea-level`
const ρₒ=1026.0 # kg m⁻³, average density at the surface of the world ocean`
Qᵘ=-ρₐ/ρₒ*cᴰʷ*u₁₀*abs(u₁₀) # m² s⁻²`
z₀ = 0.01 # m (roughness length)`
κ = 0.4 # von Karman constant`
z₁ = -0.5*znodes(Center,grid)[grid.Nz] # Closest grid center to the bottom`
cᴰᵇ = (κ / log(z₁ / z₀))^2 # Drag coefficient`

@inline drag_u(x, y, t, u, v, p) = - p.cᴰᵇ * √(u^2 + v^2) * (u)`
@inline drag_v(x, y, t, u, v, p) = - p.cᴰᵇ * √(u^2 + v^2) * (v)`
```

Here is my boundary condition, the resolved momentum flux is calculated by horizontal averaged u'w'=(u-U)*(w-W), U and W are the horizontal averaged velocity field, and the sgs flux τ₁₃ is calculated by `νₑ_∂z_u=Field(Average(∂z(u)*νₑ,dims=(1,2)))` . The total momentum flux at the boundary should be equal to my boundary condition in theory. So, when I add up the u'w' and νₑ_∂z_u, What I expect to see is that at the boundary it equals the boundary condition. (For example, at the top boundary, (u'w'+τ₁₃)/u*² should be equal to 1).
In this plot, I interpolate u'w'(an 1×1×64 Center Field) to the Z face(an 1×1×65 zFace field, the value of the first and last grids is zero) so that I can add up u'w' and τ₁₃ together.
![微信图片_20230407092538](https://user-images.githubusercontent.com/117068415/230521677-8b41573d-a591-4504-accf-494694e1c4ec.png)
Here is the result of u'w'+τ₁₃, values at the boundary are zero.
![QQ截图20230407093736](https://user-images.githubusercontent.com/117068415/230523175-29037d5b-eb84-47ba-a804-74f8a68b72d3.png)
However, when I use Qᵘ to replace the top bounded data, the result is "unsmooth". So I wonder if I can simply replace the momentum flux at the boundary with my flux boundary conditions? Or maybe there is some other way?(the green dashed line is what I expected)
![QQ截图20230407095426](https://user-images.githubusercontent.com/117068415/230525078-fc428747-43ca-4002-a67a-6e551f0922a6.png)

Thanks!!

**Reply by glwagner** [reply]
Thank you!

Just so you know, you can use triple ticks (```) to format _blocks_ of code. I took the liberty of doing that above. It really helps us read and understand your code and leads to faster solutions for you! Here's a bit about that: https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Reply by glwagner** [reply]
There are two issues. First and foremost is probably that you are treating the momentum flux `u'w'` as located at `Center, Center, Center`. This field should be located at `Face, Center, Face`:

```julia
uw_op = @at (Face, Center, Face) (u - U) * (w - W)
uw_avg = Average(uw_op, dims=(1, 2))
```

The same is true for the SGS flux.

It is correct to think of the boundary condition as the top-most grid point --- in this case, located at the cell interface `k = Nz+1`. Note that if there are `Nz` cells, then there are `Nz+1` cell interfaces in a `Bounded` direction.

Finally, if the result still is not smooth, you may want to increase vertical resolution near the top boundary if having a smooth momentum flux there is your goal.

What closure and advection scheme are you using?

**Reply by Tinydog8** [reply]
Sorry for the delay! 
The closure is AMD and the advection scheme is WENO5. I have tried to improve the resolution but the values at the second grid are still too small. Maybe there are other errors in my code. 
Thanks a lot!

**Reply by glwagner** [reply]
Did you fix the problem with the staggered location?

**Reply by Tinydog8** [reply]
Yes, they are located at (Face, Center, Face) before the average operation now.

**Reply by Tinydog8** [reply]
```julia
u = model.velocities.u
v = model.velocities.v
w = model.velocities.w

νₑ = model.diffusivity_fields.νₑ
b = model.tracers.b

U = Field(Average(u, dims=(1, 2)))
V = Field(Average(v, dims=(1, 2)))
W = Field(Average(w, dims=(1, 2)))


momentum_flux_uw_op = @at (Face,Center,Face) (u-U)*(w-W)
momentum_flux_uw = Average(momentum_flux_uw_op,dims=(1, 2))
sgs_flux_uw_op = @at (Face,Center,Face) -νₑ*∂z(u)
sgs_flux_uw = Average(sgs_flux_uw_op,dims=(1, 2))

simulation.output_writers[:averages] =
    JLD2OutputWriter(model, (; U, V,momentum_flux_uw,sgs_flux_uw),
                     schedule = AveragedTimeInterval(output_interval, window=2minutes),
                     filename = "langmuir_turbulence_averages.jld2",
                     overwrite_existing = true)


run!(simulation)

```

I calculate the momentum flux of the Langmuir Turbulence case in Oceananigans, the result is shown below. Do I have any incorrect operations here? (I have added the boundary condition to the top-most grid point here) 

Thanks a lot!

![1](https://user-images.githubusercontent.com/117068415/231475175-09640b60-3fe7-4e7f-a2a7-b5d2ab426058.png)

**Reply by glwagner** [reply]
Your code looks right --- though note that `W` is probably zero, so you don't need to write `w - W`. If you do need to subtract a horizontal average from `w`, then your domain must be heterogeneous and `W = Field(Average(w, dims=(1, 2)))` wouldn't be correct anyways (you'd need to average over only part of the domain to define a decomposition of `w`).

However, your resolution looks coarse, perhaps 2m at the top? (You haven't included the code that makes the grid.) You probably want to be 0.5m or smaller. You might try using a grid stretched towards the top if the surface matters to you.

One subtle point is that this momentum flux calculation is an approximation if you are using a high-order advection scheme. As the resolution converges I don't think this would be an issue, but maybe it matters here.

We could try averaging the "true" momentum flux. It's a bit of work but possible.

**Reply by glwagner** [reply]
You could try this:

```julia
using Oceananigans.Advection: advective_momentum_flux_Wu

wu_op = KernelFunctionOperation{Face, Center, Face}(advective_momentum_flux_Wu, grid, model.advection, w, u)
wu = Field(Average(wu_op, dims=(1, 2))
```

I believe this returns $\overline{w'u'}$ --- since $\overline{w} = 0$, we have $\overline{w u} = \overline{w' (U + u')} = \overline{w' u'}$.

**Comment by glwagner** [comment]
PS I reformatted your original post to use block quotes for the code which fixed the line spacing

**Comment by ankitsb** [comment]
Hi @Tinydog8 and @glwagner,

Very interesting discussion. Any update on this?
